From 71d4bbf7e9a5775039ce9aa0f4e6ceacdeb266fd Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Sat, 13 Dec 2014 03:06:05 +0100
Subject: [PATCH] BungeeCord Chat API


diff --git a/src/main/java/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java b/src/main/java/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java
index b5a12f0de..03a73f00f 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java
@@ -5,7 +5,17 @@ import net.minecraft.network.PacketDataSerializer;
 import net.minecraft.network.chat.IChatBaseComponent;
 import net.minecraft.network.protocol.Packet;
 
-public record ClientboundSystemChatPacket(IChatBaseComponent content, boolean overlay) implements Packet<PacketListenerPlayOut> {
+// Spigot start
+public record ClientboundSystemChatPacket(String content, boolean overlay) implements Packet<PacketListenerPlayOut> {
+
+    public ClientboundSystemChatPacket(IChatBaseComponent content, boolean overlay) {
+        this(IChatBaseComponent.ChatSerializer.toJson(content), overlay);
+    }
+
+    public ClientboundSystemChatPacket(net.md_5.bungee.api.chat.BaseComponent[] content, boolean overlay) {
+        this(net.md_5.bungee.chat.ComponentSerializer.toString(content), overlay);
+    }
+    // Spigot end
 
     public ClientboundSystemChatPacket(PacketDataSerializer packetdataserializer) {
         this(packetdataserializer.readComponent(), packetdataserializer.readBoolean());
@@ -13,7 +23,7 @@ public record ClientboundSystemChatPacket(IChatBaseComponent content, boolean ov
 
     @Override
     public void write(PacketDataSerializer packetdataserializer) {
-        packetdataserializer.writeComponent(this.content);
+        packetdataserializer.writeUtf(this.content, 262144); // Spigot
         packetdataserializer.writeBoolean(this.overlay);
     }
 
-- 
2.37.0

